### PR TITLE
Alphabetized all .dmi files in icons\effects

### PR DESCRIPTION
**Pull request by NIXC**
I have created a python program that automagically alphabetizes .dmi files. I figured I would run it on just a small portion before doing the entire icons folder just to see how interested everyone was.
What I did:
1. Ran the alphabetizer on icons\effects
2. Opened and saved every file in icons\effects to ensure the proper file structure (its hard to do a complete clone and my png library has a bug in saving paletted files that I'm too lazy to fix)
3. Ran around for a bit. Nothing obvious was using the wrong icon (I even nuked the station).

I havent had any issues just running the alphabetizer on the whole icons dir without saving, so I figure this should be safe. I did some checks to ensure things were still in the right places.
This makes the more full icon files easier to read(especially some of the item dmi's), and should help forks rebase easier. If there are no objections I will make another PR with more/the rest of the icons.
I will also be releasing the tool when I make the cmd line interface less dumb.

This makes no player visible changes

### CODE REVIEW

**Comment by NIXC** [comment]
While I'm thinking of it I'm going to update icons\obj\atmospherics\pipes\large.dmi to dmi v4 because it annoys me. That's just opening and resaving the file.

**Comment by optimumtact** [comment]
@KorPhaeron @ChangelingRain @lzimann test merge this please

**Comment by ChangelingRain** [comment]
I'm not sure this is a good idea from the dev side of things because it removes any pre-existing orders in the files and that can just make things worse.
Like, seriously, if you run this on the inhand files, which are already bad and difficult to find stuff in, and remove existing ordering in favor of alphabetical order will very definitely make that worse, especially with all our wonderful badly-named icon states.

**Comment by NIXC** [comment]
While I started with alphabetical in mind, its a simple python lambda/key function away from sorting by any other naming convention. Or we can skip files, if their preexisting sort is better.

**Comment by ChangelingRain** [comment]
I'm going to look through all the files and give you a list of ones I think would be improved.

**Comment by ChangelingRain** [comment]
`human_face.dmi`
`status_display.dmi` assuming it takes capitalization into account

**Comment by NIXC** [comment]
Hmm. So your saying That out of all the files only those 2 should be alphabetized? I know there were a few mob ones that gave me a heart attack the other day. It does take capitalization into account (I think its ready for the lime light now, Im going to throw it in the tools folder in a seperate PR)

Can you think of any other form of sorting that may be better suited for other files? Some of them are a real mess.

**Comment by ChangelingRain** [comment]
the issue with sorting is that usually each file wants to be sorted by what the icon IS, and not what it's actually named.
there are some exception that could probably have this applied TO them with the right rule, but it's not like they're going to get randomly shuffled so it's not super important to do so.

**Comment by NIXC** [comment]
It seems to be sorted by when it was added to the game right now for most files though.

**Comment by KorPhaeron** [comment]
Good change

**Comment by ChangelingRain** [comment]
Yes, and that apparently by-date sort has a secondary human sort on top of it; icons similar in theme or used for the same feature tend to get added next to similarly-themed icons and icons for the same feature, even if they're not in alphabetical order.

Take, as example, `actions.dmi`, which, while mostly by date, has a few sub-orders for stuff like mechs which have had things added by the above 'human' sort rather than by date.

**Comment by NIXC** [comment]
So any word on this? Should I switch to instead cherry picking a few files to sort? Should I come up with some form of manual sorting method?
 I think im finally happy with my program so Ill probably PR it to the tools folder tonight.

**Comment by lzimann** [comment]
I have no problems with alphabetically sorting every file, sometimes it's a pain to search for icons(well i wrote a tool to do that for me, so it doesn't matter) when there's no obvious order on what's going on.

**Comment by ChangelingRain** [comment]
I mean it might be a useful tool, but I'd rather you not immediately do it to every file in the effects folder.

**Comment by NIXC** [comment]
Ok, i will check over the files and remove a few like static. If anyone else would like to test this i would be happy to hear if anything breaks(because it really shouldn't)
I'm going to see about making some form of algorithm that tries to decide how it would best be to sort the file, dont know how far i will get.

**Comment by KorPhaeron** [comment]
Should just do it to every dmi

At least this way if I dont know an icon name I can look up its object and then quickly find it vs hitting the first letter and jumping around a huge file at random

**Comment by NIXC** [comment]
My ultimate dream is that we can just run this on the whole icon file every time changes are made and have everything remain decently organized

**Comment by optimumtact** [comment]
I say we vote to ignore @ChangelingRain and just do every file

up thumbs my pr if you agree, down thumbs if you dont, explode if you don't care

**Comment by NIXC** [comment]
That would appear to be a vote from everyone who has participated in this thread. Doesnt seem to be any other naysay besides Changeling Rain.

**Comment by duncathan** [comment]
I would prefer if we at least cherry picked icon files to leave as is, and then alphabetize the rest

**Comment by KorPhaeron** [comment]
@duncathan 

What is the advantage to leaving things as is? Short term it might make life slightly easier for current coders who happen to have memorized a particular file but long term it I think it will be best if everything follows rules that anyone can grasp

File: tools/dmiAlphabetizer/alphabetsort.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    ztxt = img.info["Description"]
+    new_data = [] #Last time this is going to look nice
+    ztxt = ztxt.split('\n') #UNFUCKED
+    pos = 4 # we start at 4 cuz the first few are hardcoded
+    width = 0
+    height = 0
+    force = 0
+    if ztxt[1].split()[2] != "4.0":
+        print("WARNING: VERSION CHANGE DETECTED")
+        print("Shit could be fucked.")
```
Comment (ike709) [inline tools/dmiAlphabetizer/alphabetsort.py]: 👍

**Comment by NIXC** [comment]
I have added the program I used to do the actual alphabetization. Feel free to have a look at it and pick it apart. (I have a solo PR of it on FTL13 repo)

**Comment by lzimann** [comment]
I really don't see a reason to not just do this in every icon file.

**Comment by MrStonedOne** [comment]
Thats because you didn't read this thread where reasons were already given.

File: tools/dmiAlphabetizer/readme [hunk 2]
```diff
@@ -0,0 +1,17 @@
+So for this thing to work you need to install pillow for python 3. any flavor
```
Comment (NIXC) [inline tools/dmiAlphabetizer/readme]: What does MD stand for?

**Comment by lzimann** [comment]
Let me rephrase then, a good reason(in my opinion anyway).

File: tools/dmiAlphabetizer/readme (hunk 2 above)
Comment (lzimann) [inline tools/dmiAlphabetizer/readme]: Markdown

File: tools/dmiAlphabetizer/alphabetsort.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
+#NINX3's dank .dmi file sorter.
+import PIL #pep install pillow
```
Comment (NIXC) [inline tools/dmiAlphabetizer/alphabetsort.py]: The requirements are also covered in the readme.md, should I add another file to rehash it again? (I cant find any examples of a requirements.txt, so I dont know if its readme style or some form of automagic system)
The worst that happens without it is the program crashes and you (hopefully?) read the comment.

**Comment by MrStonedOne** [comment]
To add:

Similar icons near each other compresses better in png, not all similar icons have similar names.

**Comment by MrStonedOne** [comment]
I feel like it would be better to just nag byond to add a sort option to the display order of the dmi, or make a new editor.

File: tools/dmiAlphabetizer/alphabetsort.py (hunk 3 above)
Comment (MrStonedOne) [inline tools/dmiAlphabetizer/alphabetsort.py]: This does not specify what python version it requires

File: tools/dmiAlphabetizer/readme.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+it needs to be resaved so that the metadata is what byond expects. Now I know that it will run
+without resaving (byond's current version doesn't seem to care) but future versions may not.
+And for small files the file is usually bigger (especially paletted pngs, since they have a bug
+in the Pillow library afaik, or I'm incapable of the mental gymnastics to fix it) before you resave.
+also paths like ..\..\icons should work fine when used with the recurser while on windows.
+Mileage on other operating systems may vary.
+##Requirements
+So for this thing to work you need to install pillow for python 3. any flavor
+of python 3 should work. Ensure you don't have PIL installed (you should know already
+if you do).
```
Comment (MrStonedOne) [inline tools/dmiAlphabetizer/readme.md]: Requirements should be a easy to read format like a list or bullet points.

Requirements in complete sentences like this are hard to read and parse, especially when you come back after fulfilling one to see the next one.

**Comment by NIXC** [comment]
I dont feel we need to go that crazy about compression, It wouldnt save us anything in the long run. The entire icon file is 35mb, and I cant see sorting it in order to reduce size causing a huge change.
Nagging Byond to add native support to sorting would be amazing, but I dont know how fast that would happen.

File: tools/dmiAlphabetizer/readme.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+it needs to be resaved so that the metadata is what byond expects. Now I know that it will run
+without resaving (byond's current version doesn't seem to care) but future versions may not.
+And for small files the file is usually bigger (especially paletted pngs, since they have a bug
+in the Pillow library afaik, or I'm incapable of the mental gymnastics to fix it) before you resave.
+also paths like ..\..\icons should work fine when used with the recurser while on windows.
+Mileage on other operating systems may vary.
+## Requirements
+So for this thing to work you need to install pillow for python 3. any flavor
+of python 3 should work. Ensure you don't have PIL installed (you should know already
+if you do).
```
Comment (MrStonedOne) [inline tools/dmiAlphabetizer/readme.md]: Requirements should be a easy to read format like a list or bullet points.

Requirements in complete sentences like this are hard to read and parse, especially when you come back after fulfilling one to see the next one.

**Comment by optimumtact** [comment]
if you were concerned about dmi size you would be running optipng mso, why the sudden concern for compression?

as an aside this would also be solved by giving common items common prefixes.

We should create an icon naming standard

**Comment by ChangelingRain** [comment]
Nagging byond for a sort option would certainly be much more practical and useful, honestly. I don't want the order to get fucked up because I tend to rely on it.

**Comment by NIXC** [comment]
@ChangelingRain How can the current semi random ordering in a file be better than having to remember a new alphabetical ordering of a file? It may be easier for you, but would it be easier for someone new to the code base?

**Comment by ChangelingRain** [comment]
thus, I want both.

**Comment by optimumtact** [comment]
Lummox has already made it clear he does not have any intent to improve the dream editor, so waiting for byond is not a valid path of action.

I say we go ahead with this.

**Comment by duncathan** [comment]
i changed my mind, this is ok

**Comment by optimumtact** [comment]
@Cheridan this is split fairly evenly amongst maintainers, so this is probably best decided by you.

**Comment by Cheridan** [comment]
I agree with Joan. If you use ammo.dmi as an example, all of the shotgun shells (cshell, ishell, mshell, etc.) are next to each other, which makes sense. If you arranged everything alphabetically it'd suddenly be totally out of order. 

In theory it would make things easier to find, but most of our files aren't large enough for finding a certain sprite to be an issue. Most times, the difficulty is finding the correct .dmi file, not finding the sprite within it.

**Comment by NIXC** [comment]
Rip. Would I have support of cherry picking a few of the larger files that would benefit from this, or is it not worth the time?

**Comment by ChangelingRain** [comment]
I mean, you could suggest some but most don't benefit and all, and stuff like inhand files would be better done by splitting them up
